### PR TITLE
Add application image

### PR DESCRIPTION
**Pull request by cfajardo**
## Description

Application picture feature added.

### Screenshot

image

image

### Issues Fixed or Closed

N/A

## Type of Change

- Application image path into *_apps.json.
- Modification of GFE appimage route to send the picture to the client.
- A field has been added to the WebUI to configure image path.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the documentation blocks for new or existing components

### CODE REVIEW

File: sunshine/process.cpp [hunk 1]
```diff
@@ -279,6 +307,7 @@ std::optional<proc::proc_t> parse(const std::string &file_name) {
       auto output             = app_node.get_optional<std::string>("output"s);
       auto name               = parse_env_val(this_env, app_node.get<std::string>("name"s));
       auto cmd                = app_node.get_optional<std::string>("cmd"s);
+      auto image_path              = app_node.get_optional<std::string>("image-path"s);
```
Comment (ReenigneArcher) [inline sunshine/process.cpp:310]: The "=" symbols should be aligned.

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
Code reviewed, but not tested. Seems mostly okay at first glance. One small change requested.

Has you tested on Linux and Windows? Please check each box if you've tested.
- [ ] Windows
- [ ] Linux

File: sunshine/process.cpp (hunk 1 above)
Comment (cfajardo) [reply sunshine/process.cpp]: done

**Comment by cfajardo** [comment]
> 

[x] Windows
[ ] Linux

I don't have a linux setup now. Currently building one.

**Review by ReenigneArcher — APPROVED** [review]

**Comment by TheElixZammuto** [comment]
Tested with Windows 11 and Moonlight-qt and Moonlight-ios with both JPG and PNG files, it works

**Comment by cgutman** [comment]
I'd be careful returning non-PNG images. They might work but it's only by luck.

All current Moonlight clients assume the image they get back will be a PNG. They will cache the images locally as .png files, which may confuse the APIs consuming those files if they guess the image format based on the file extension.

File: sunshine/process.cpp [hunk 2]
```diff
@@ -189,6 +190,33 @@ std::vector<ctx_t> &proc_t::get_apps() {
...
+  auto image_extention = std::filesystem::path(app_image_path).extension().string();
+  image_extention = image_extention.substr(1, image_extention.length() - 1);
+
+  std::error_code code;
+  if (!std::filesystem::exists(app_image_path, code) || !CHECK_EXPECTED_PICTURE_EXTENTIONS(image_extention)) {
+    return { SUNSHINE_ASSETS_DIR "/box.png", "png" };
+  }
+
+  // return only "content-type" http header compatible image type.
+  return { app_image_path, image_extention == "jpg" ? "jpeg" : image_extention };
```
Comment (ReenigneArcher) [inline sunshine/process.cpp:217]:
```cpp
return { app_image_path, image_extension };
```

Typos here + limit to PNG format

File: sunshine/process.cpp [hunk 3]
```diff
@@ -189,6 +190,33 @@ std::vector<ctx_t> &proc_t::get_apps() {
...
+  auto app_image_path = _apps[app_index].image_path;
+  if (app_image_path.empty()) {
+    return { SUNSHINE_ASSETS_DIR "/box.png", "png" };
+  }
+
+  auto image_extention = std::filesystem::path(app_image_path).extension().string();
+  image_extention = image_extention.substr(1, image_extention.length() - 1);
+
+  std::error_code code;
+  if (!std::filesystem::exists(app_image_path, code) || !CHECK_EXPECTED_PICTURE_EXTENTIONS(image_extention)) {
```
Comment (ReenigneArcher) [inline sunshine/process.cpp:212]:
```cpp
if (!std::filesystem::exists(app_image_path, code) || !CHECK_EXPECTED_PICTURE_EXTENSIONS(image_extension)) {
```

Typos here

File: assets/web/apps.html [hunk 4]
```diff
@@ -166,6 +166,22 @@ <h1>Applications</h1>
...
+          class="form-control monospace"
+          id="appImagePath"
+          aria-describedby="appImagePathHelp"
+          v-model="editForm['image-path']"
+        />
+        <div id="appImagePathHelp" class="form-text">
+          Application icon/picture/image path that will be sent to client.
+          Only full path are working, so no relative path.
+          If not set, Sunshine will send default box image.
+        </div>
```
Comment (ReenigneArcher) [inline assets/web/apps.html:183]:
```html
        <div id="appImagePathHelp" class="form-text">
          Application icon/picture/image path that will be sent to client.
          Only full path are working, so no relative path. Image must be a PNG.
          If not set, Sunshine will send default box image.
        </div>
```

Something along these lines... Add a note to indicate the image needs to be a png.

File: sunshine/process.cpp [hunk 5]
```diff
@@ -189,6 +190,33 @@ std::vector<ctx_t> &proc_t::get_apps() {
...
+    return { SUNSHINE_ASSETS_DIR "/box.png", "png" };
+  }
+
+  auto app_image_path = _apps[app_index].image_path;
+  if (app_image_path.empty()) {
+    return { SUNSHINE_ASSETS_DIR "/box.png", "png" };
+  }
+
+  auto image_extention = std::filesystem::path(app_image_path).extension().string();
+  image_extention = image_extention.substr(1, image_extention.length() - 1);
```
Comment (ReenigneArcher) [inline sunshine/process.cpp:209]:
```cpp
auto image_extension = std::filesystem::path(app_image_path).extension().string();
image_extension = image_extension.substr(1, image_extension.length() - 1);
```

Typos

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
Based on comments by @cgutman let's force the image to be PNG.

Additionally there was a typo in the word `extension` wherever it appeared.

File: sunshine/process.h [hunk 6]
```diff
@@ -16,6 +16,8 @@
 
 #include "utility.h"
 
+#define CHECK_EXPECTED_PICTURE_EXTENTIONS(extention) (extention == "png" || extention == "jpg"  || extention == "jpeg")
```
[the remark this answers is not in the document]
Comment (cfajardo) [reply sunshine/process.h:19]: I will remove this macro. As only png is supported by GFE api, Sunshine should stick to same specs, then a macro only obfuscate the code.

File: sunshine/process.cpp (hunk 2 above)
Comment (cfajardo) [reply sunshine/process.cpp:217]: This method will only return a valid file path. Only png files will be returned.

I don't like having code that may be used. The future change, if Moonlight supports jpg, is pretty simple.

File: sunshine/process.cpp (hunk 3 above)
Comment (cfajardo) [reply sunshine/process.cpp:212]: macro will be removed.

File: assets/web/apps.html (hunk 4 above)
Comment (cfajardo) [reply assets/web/apps.html:183]: done

File: sunshine/process.cpp (hunk 5 above)
Comment (cfajardo) [reply sunshine/process.cpp]: done

File: sunshine/process.cpp [hunk 7]
```diff
@@ -193,28 +193,28 @@ std::vector<ctx_t> &proc_t::get_apps() {
...
 
-  auto image_extention = std::filesystem::path(app_image_path).extension().string();
-  image_extention = image_extention.substr(1, image_extention.length() - 1);
+  auto image_extension = std::filesystem::path(app_image_path).extension().string();
+  image_extension = image_extension.substr(1, image_extension.length() - 1);
 
   std::error_code code;
-  if (!std::filesystem::exists(app_image_path, code) || !CHECK_EXPECTED_PICTURE_EXTENTIONS(image_extention)) {
-    return { SUNSHINE_ASSETS_DIR "/box.png", "png" };
+  if (!std::filesystem::exists(app_image_path, code) || image_extension != "png") {
```
Comment (ReenigneArcher) [inline sunshine/process.cpp:212]: Question as I am getting back up to speed with c++... does `std::filesystem::path::extension` account for capitalization? Maybe a good idea to convert image_extension to lowercase?

**Review by ReenigneArcher — COMMENTED** [review]
Reviewed.

File: sunshine/process.cpp (hunk 7 above)
Comment (cfajardo) [reply sunshine/process.cpp:213]: You're right. It's done.

**Review by ReenigneArcher — APPROVED** [review]

**Comment by ReenigneArcher** [comment]
@cfajardo can you take a look at this? A user in the moonlight discord was testing and couldn't get it to return the png or even the default image (on Windows). https://discord.com/channels/352065098472488960/667129596088418344/938115637627191417

I'm also going to try to setup a test machine later to test out this PR in both linux and windows.

**Comment by cfajardo** [comment]
> @cfajardo can you take a look at this? A user in the moonlight discord was testing and couldn't get it to return the png or even the default image (on Windows). https://discord.com/channels/352065098472488960/667129596088418344/938115637627191417
> 
> I'm also going to try to setup a test machine later to test out this PR in both linux and windows.

Could you invite me to Discord channel ? I'm not able to access it.

**Comment by ReenigneArcher** [comment]
> > @cfajardo can you take a look at this? A user in the moonlight discord was testing and couldn't get it to return the png or even the default image (on Windows). https://discord.com/channels/352065098472488960/667129596088418344/938115637627191417
> > I'm also going to try to setup a test machine later to test out this PR in both linux and windows.
> 
> Could you invite me to Discord channel ? I'm not able to access it.

https://moonlight-stream.org/discord

**Comment by cfajardo** [comment]
> > > @cfajardo can you take a look at this? A user in the moonlight discord was testing and couldn't get it to return the png or even the default image (on Windows). https://discord.com/channels/352065098472488960/667129596088418344/938115637627191417
> > > I'm also going to try to setup a test machine later to test out this PR in both linux and windows.
> > 
> > 
> > Could you invite me to Discord channel ? I'm not able to access it.
> 
> https://moonlight-stream.org/discord

Sunshine channel seems to be private. Am I right ?

**Comment by ReenigneArcher** [comment]
@cfajardo No... I think you need to select your roles in https://discord.com/channels/352065098472488960/668086277639503874

Also, I can confirm I'm not seeing images on a windows host, haven't tested linux yet.

File: assets/apps_linux.json [hunk 8]
```diff
@@ -13,7 +13,8 @@
 		"name":"Steam BigPicture",
 
 		"output":"steam.txt",
-		"detached":["setsid steam steam://open/bigpicture"]
+		"detached":["setsid steam steam://open/bigpicture"],
+		"image-path":"./asset/steam.png"
```
Comment (ReenigneArcher) [inline assets/apps_linux.json:17]:
```cpp
"image-path":"./assets/steam.png"
```

**Review by ReenigneArcher — COMMENTED** [review]
Small typo in apps_linux.json

File: assets/apps_linux.json [hunk 9]
```diff
@@ -13,7 +13,8 @@
 		"name":"Steam BigPicture",
 
 		"output":"steam.txt",
-		"detached":["setsid steam steam://open/bigpicture"]
+		"detached":["setsid steam steam://open/bigpicture"],
+		"image-path":"./assets/steam.png"
```
Comment (TheElixZammuto) [inline assets/apps_linux.json:17]: Are we sure we can actually use the steam image? Steam is a trademarked logo after all

**Review by TheElixZammuto — CHANGES_REQUESTED** [review]

File: assets/apps_linux.json (hunk 9 above)
Comment (ReenigneArcher) [reply assets/apps_linux.json:17]: Ironically, the Steam artwork used by GFE does not follow their guidelines, haha.

Perhaps, the colors need to be adjusted a bit. Seems they want specific colors. Although I doubt they would throw a fit over this. Below is the colors they use in the guidelines.

Also we need to add this attribution somewhere... maybe in the webui?

```txt
©2018 Valve Corporation. Steam and the Steam logo are trademarks and/or
registered trademarks of Valve Corporation in the U.S. and/or other countries. All
rights reserved.
```
image

File: assets/web/apps.html [hunk 10]
```diff
@@ -166,6 +166,22 @@ <h1>Applications</h1>
...
       </div>
+      <!-- Image path -->
+      <div class="mb-3">
+        <label for="appImagePath" class="form-label">Image</label>
+        <input
+          type="text"
+          class="form-control monospace"
+          id="appImagePath"
+          aria-describedby="appImagePathHelp"
+          v-model="editForm['image-path']"
```
[the remark this answers is not in the document]
Comment (cfajardo) [reply assets/web/apps.html:177]: Thanks for the review.
Can you give more information so I can reproduce it ? I've just tested with a path with white spaces and the path is saved properly.

[the remark this answers is not in the document]
Comment (cfajardo) [reply assets/web/apps.html:177]: Ok, just understand what you've done. You paste the path with double quote. Ok I'll do the modification.

File: assets/apps_linux.json (hunk 9 above)
Comment (TheElixZammuto) [reply assets/apps_linux.json:17]: We can simply add this onto a NOTICE file on the root of the repo and then bundle it inside the zip file on the release, we should be okay then

File: assets/web/apps.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (cfajardo) [reply assets/web/apps.html:177]: Done and I've done some tests. Seems to be ok. Fine for you too ?

File: assets/apps_linux.json (hunk 9 above)
Comment (cfajardo) [reply assets/apps_linux.json:17]: ok

File: assets/web/apps.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (ReenigneArcher) [reply assets/web/apps.html:177]: It's a good idea to automatically remove the double quotes if they're present. Will save some issues down the road.

File: assets/apps_linux.json (hunk 9 above)
Comment (cfajardo) [reply assets/apps_linux.json:17]: Done. CMake also copy the file into build directory, but don't know if it's useful.
I've some questions :
- Linux packages : Where do we put the file ?
- Is there other package to modify ?

File: assets/web/apps.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheElixZammuto) [reply assets/web/apps.html:177]: I haven't tested but from the code it seems okay

File: assets/apps_linux.json (hunk 9 above)
Comment (TheElixZammuto) [reply assets/apps_linux.json:17]: Packaging is done by the GitHub actions, I don't remember the exact details but it definitely can be edited to include the file in the artifacts (I don't know if it should be done here or in another PR though)

**Review by TheElixZammuto — APPROVED** [review]

File: assets/apps_linux.json (hunk 9 above)
Comment (ReenigneArcher) [reply assets/apps_linux.json:17]: Let's take care of it in another PR, probably would be good to add the LICENSE too.

File: assets/web/apps.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (ReenigneArcher) [reply assets/web/apps.html:177]: I tested on Windows. Using with and without quotes is working.

File: CMakeLists.txt [hunk 11]
```diff
@@ -364,6 +364,7 @@ endif()
 list(APPEND SUNSHINE_DEFINITIONS SUNSHINE_ASSETS_DIR="${SUNSHINE_ASSETS_DIR}")
 list(APPEND SUNSHINE_DEFINITIONS SUNSHINE_CONFIG_DIR="${SUNSHINE_CONFIG_DIR}")
 list(APPEND SUNSHINE_DEFINITIONS SUNSHINE_DEFAULT_DIR="${SUNSHINE_DEFAULT_DIR}")
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/NOTICE" "NOTICE")
```
Comment (ReenigneArcher) [inline CMakeLists.txt:367]: One last change, undo adding this line. We are going to add the notice to the web UI (separate PR for that). After this change I will merge this PR. Thanks!

**Review by ReenigneArcher — COMMENTED** [review]
I appreciate your work on this PR and apologize for asking for a bunch of changes. One final change and then I'll merge the PR.

File: CMakeLists.txt (hunk 11 above)
Comment (cfajardo) [reply CMakeLists.txt]: Done

**Review by ReenigneArcher — APPROVED** [review]

**Pull request merged**